packages/multichain/src/middlewares/MultichainSubscriptionManager.ts

```diff
@@ -0,0 +1,162 @@
+import { toHex } from '@metamask/controller-utils';
+import createSubscriptionManager from '@metamask/eth-json-rpc-filters/subscriptionManager';
+import type { NetworkController } from '@metamask/network-controller';
+import SafeEventEmitter from '@metamask/safe-event-emitter';
+import type { CaipChainId, Hex } from '@metamask/utils';
+import { parseCaipChainId } from '@metamask/utils';
+
+import type { ExternalScopeString } from '../scope/types';
+import type { ExtendedJsonRpcMiddleware } from './MultichainMiddlewareManager';
+
+export type SubscriptionManager = {
+  events: SafeEventEmitter;
+  destroy?: () => void;
+  middleware: ExtendedJsonRpcMiddleware;
+};
+
+type SubscriptionNotificationEvent = {
+  jsonrpc: '2.0';
+  method: 'eth_subscription';
+  params: {
+    subscription: Hex;
+    result: unknown;
+  };
+};
+
+type SubscriptionKey = {
+  scope: ExternalScopeString;
+  origin: string;
+  tabId?: number;
+};
+type SubscriptionEntry = SubscriptionKey & {
+  subscriptionManager: SubscriptionManager;
+};
+
+type MultichainSubscriptionManagerOptions = {
+  findNetworkClientIdByChainId: NetworkController['findNetworkClientIdByChainId'];
+  getNetworkClientById: NetworkController['getNetworkClientById'];
+};
+
+export class MultichainSubscriptionManager extends SafeEventEmitter {
+  #findNetworkClientIdByChainId: NetworkController['findNetworkClientIdByChainId'];
+
+  #getNetworkClientById: NetworkController['getNetworkClientById'];
+
+  #subscriptions: SubscriptionEntry[] = [];
+
+  constructor(options: MultichainSubscriptionManagerOptions) {
+    super();
+    this.#findNetworkClientIdByChainId = options.findNetworkClientIdByChainId;
+    this.#getNetworkClientById = options.getNetworkClientById;
+  }
+
+  onNotification(
+    { scope, origin, tabId }: SubscriptionKey,
+    { method, params }: SubscriptionNotificationEvent,
+  ) {
+    this.emit('notification', origin, tabId, {
```

Comment:
I would expect a method prefixed with `on` to subscribe to an event, not emit one.

What do you think about calling this method `notify`?

Reply:
Yes! https://github.com/MetaMask/core/pull/4813/commits/ff57ba39b51692f2b39e94aac021cedba77a3720